File: cane-csharp/cane-core/Runtime/PlaybackControls.cs
using System;
using Cane.Animation;

namespace Cane
{
    public sealed partial class RuntimePlayer
    {
        private RuntimeFrame ConfigureCurrent(string operation, int track, Action<Entry> change)
        {
            RequireIdle(operation); long identity = Current(playback, track, operation).EventOrder;
            return Configure(operation, (s, _) => change(Current(s, track, operation)),
                configureBaseline: s => change(BaselineEntry(s, track, identity, operation)));
        }
        private static Entry BaselineEntry(Playback state, int track, long identity, string operation)
        {
            if (state.Tracks.TryGetValue(track, out Track? t))
            {
                for (Entry? e = t.Current; e != null; e = e.From) if (e.EventOrder == identity) return e;
                foreach (Entry e in t.Queue) if (e.EventOrder == identity) return e;
            }
            throw new RuntimeException(RuntimeErrorCode.InvalidState, operation, "The playback entry is missing from its configuration baseline.", "trackIndex");
        }
        private static Entry Queued(Playback state, int track, int queueIndex, string operation)
        {
            if (queueIndex < 0) throw new RuntimeException(RuntimeErrorCode.InvalidArgument, operation, "Queue index must be non-negative.", "queueIndex");
            if (!state.Tracks.TryGetValue(track, out Track? t) || queueIndex >= t.Queue.Count)
                throw new RuntimeException(RuntimeErrorCode.NotFound, operation, "Queued entry is missing.", "queueIndex");
            return t.Queue[queueIndex];
        }
        public RuntimeFrame SetAnimationTime(float timeSeconds)
        {
            const string op = "setAnimationTime"; Nonnegative(timeSeconds, op, "timeSeconds"); RequireIdle(op);
            long? identity = playback.Tracks.TryGetValue(0, out Track? t) ? t.Current.EventOrder : (long?)null;
            void Set(Playback state, bool origin)
            {
                state.DetachedTime = timeSeconds;
                if (!identity.HasValue) return;
                Entry e = origin ? BaselineEntry(state, 0, identity.Value, op) : Current(state, 0, op);
                e.Time = Numeric.Clamp(timeSeconds, e.Start, e.End) - e.Start; e.CursorInitialized = true;
            }
            return Configure(op, (s, _) => Set(s, false), configureBaseline: s => Set(s, true));
        }
        public RuntimeFrame SetQueuedEntryOptions(int trackIndex, int queueIndex, QueuedEntryOptions options)
        {
            const string op = "setQueuedEntryOptions"; Index(trackIndex, op); RequireIdle(op);
            if (options == null) throw new RuntimeException(RuntimeErrorCode.InvalidArgument, op, "Queued entry options are required.", "options");
            float? delay = options.DelaySeconds, mix = options.MixDurationSeconds; bool? loop = options.Looping;
            if (!delay.HasValue && !mix.HasValue && !loop.HasValue) throw new RuntimeException(RuntimeErrorCode.InvalidArgument, op, "At least one queued entry option is required.", "options");
            if (delay.HasValue) Nonnegative(delay.Value, op, "delaySeconds"); if (mix.HasValue) Nonnegative(mix.Value, op, "mixDurationSeconds");
            long identity = Queued(playback, trackIndex, queueIndex, op).EventOrder;
            void Edit(Entry entry)
            { if (delay.HasValue) entry.Delay = delay.Value; if (mix.HasValue) entry.MixDuration = mix.Value; if (loop.HasValue) entry.Loop = loop.Value; }
            return Configure(op, (s, _) => Edit(Queued(s, trackIndex, queueIndex, op)),
                configureBaseline: s => Edit(BaselineEntry(s, trackIndex, identity, op)));
        }
        public RuntimeFrame RemoveQueuedEntry(int trackIndex, int queueIndex)
        {
            const string op = "removeQueuedEntry"; Index(trackIndex, op); RequireIdle(op);
            long identity = Queued(playback, trackIndex, queueIndex, op).EventOrder;
            return Configure(op, (s, events) => {
                Entry removed = Queued(s, trackIndex, queueIndex, op); s.Tracks[trackIndex].Queue.RemoveAt(queueIndex);
                events.Lifecycle("dispose", trackIndex, removed, 0);
            }, configureBaseline: s => {
                Entry removed = BaselineEntry(s, trackIndex, identity, op);
                if (!s.Tracks[trackIndex].Queue.Remove(removed)) throw new RuntimeException(RuntimeErrorCode.InvalidState, op, "The baseline entry is no longer queued.", "queueIndex");
            });
        }
        public RuntimeStep AdvancePhysics(float deltaSeconds)
        {
            const string op = "advancePhysics"; Nonnegative(deltaSeconds, op, "deltaSeconds"); ulong sequence = NextSequence(op);
            try
            {
                Playback state = playback.Clone();
                RuntimeFrame frame = Evaluate(state, host, sequence, default, out Pose pose, operation: op, physicsDeltaOverride: deltaSeconds);
                playback = state; Frame = frame; publishedPose = pose;
                return new RuntimeStep(deltaSeconds > 0, frame, Array.Empty<RuntimeEvent>());
            }
            catch (RuntimeException e) { throw Reframe(e, op); }
        }
    }
}

File: cane-cpp/cane-core/src/player/playback_controls.cpp
#include "state.hpp"

namespace cane {
namespace {
using detail::PlayerState;
using animation::EventSink;
using animation::Playback;
using animation::TrackEntry;
using ClockMutation = std::function<void(Playback&, EventSink&)>;
void clocks(detail::RuntimePlayerImpl& p, const char* op, const ClockMutation& mutate, bool publish = true) {
    p.configure(op, [&](PlayerState& s, EventSink& live, EventSink& baseline) { mutate(s.live, live); mutate(s.baseline, baseline); }, publish);
}
const TrackEntry& current(const Playback& state, std::uint32_t index, const char* op) {
    (void)state.query(index); const auto at = state.tracks().find(index);
    if (at == state.tracks().end()) throw Error(ErrorCode::not_found, op, "Current entry does not exist.", "trackIndex");
    return at->second.chain.back();
}
void current_edit(detail::RuntimePlayerImpl& p, const char* op, std::uint32_t index,
                  const std::function<void(Playback&)>& edit, const std::function<void(TrackEntry&, const TrackEntry&)>& project) {
    p.configure(op, [&](PlayerState& s, EventSink&, EventSink&) {
        edit(s.live); const auto& e = current(s.live, index, op);
        project(s.baseline.find_entry(index, e.identity, op), e);
    });
}
}
RuntimeStep RuntimePlayer::advance_physics(float delta_seconds) {
    constexpr auto op = "advancePhysics";
    return detail::operation(op, [&] {
        auto& p = impl(); const auto sequence = p.next_sequence(op); detail::require_nonnegative(delta_seconds, op, "deltaSeconds");
        auto candidate = std::make_unique<PlayerState>(*p.state); constraints::PhysicsBudget budget;
        p.publish(*candidate, p.evaluate(*candidate, budget, {}, nullptr, op, nullptr, sequence, delta_seconds), sequence);
        RuntimeStep result{sequence, candidate->frame->time_seconds(), delta_seconds > 0, {}};
        p.state.swap(candidate); return result;
    });
}
void RuntimePlayer::set_default_mix(float seconds) { clocks(impl(), "setDefaultMix", [&](Playback& p, EventSink&) { p.set_default_mix(seconds); }, false); }
void RuntimePlayer::set_mix(std::string_view from, std::string_view to, float seconds) { clocks(impl(), "setMix", [&](Playback& p, EventSink&) { p.set_mix(from, to, seconds); }, false); }
void RuntimePlayer::set_animation(std::uint32_t track, std::string_view id, bool looping, std::optional<float> mix) {
    clocks(impl(), "setAnimation", [&](Playback& p, EventSink& sink) { p.replace(track, id, looping, mix, sink); });
}
void RuntimePlayer::queue_animation(std::uint32_t track, std::string_view id, bool looping, float delay, std::optional<float> mix) {
    clocks(impl(), "queueAnimation", [&](Playback& p, EventSink& sink) { p.enqueue(track, id, looping, delay, mix, sink); });
}
void RuntimePlayer::set_empty_animation(std::uint32_t track, float mix) { clocks(impl(), "setEmptyAnimation", [&](Playback& p, EventSink& sink) { p.replace(track, {}, false, mix, sink); }); }
void RuntimePlayer::queue_empty_animation(std::uint32_t track, float mix, float delay) { clocks(impl(), "queueEmptyAnimation", [&](Playback& p, EventSink& sink) { p.enqueue(track, {}, false, delay, mix, sink); }); }
void RuntimePlayer::clear_track(std::uint32_t track) { clocks(impl(), "clearTrack", [&](Playback& p, EventSink& sink) { p.clear(track, sink); }); }
void RuntimePlayer::clear_tracks() { clocks(impl(), "clearTracks", [](Playback& p, EventSink& sink) { p.clear_all(sink); }); }
void RuntimePlayer::set_track_time(std::uint32_t track, float time) {
    current_edit(impl(), "setTrackTime", track, [&](Playback& p) { p.set_time(track, time); },
                 [](TrackEntry& baseline, const TrackEntry& live) { baseline.time = live.time; baseline.cursor_initialized = live.cursor_initialized; });
}
void RuntimePlayer::set_animation_time(float time) {
    impl().configure("setAnimationTime", [&](PlayerState& s, EventSink&, EventSink&) {
        s.live.set_animation_time(time); s.baseline.set_detached_time(time);
        if (!s.live.query(0)) return;
        const auto& e = current(s.live, 0, "setAnimationTime"); auto& b = s.baseline.find_entry(0, e.identity, "setAnimationTime");
        b.time = e.time; b.cursor_initialized = true;
    });
}
void RuntimePlayer::set_track_options(std::uint32_t track, const TrackOptions& options) {
    current_edit(impl(), "setTrackOptions", track, [&](Playback& p) { p.set_options(track, options); }, [](TrackEntry& b, const TrackEntry& e) {
        b.alpha = e.alpha; b.rate = e.rate; b.event_threshold = e.event_threshold; b.attachment_threshold = e.attachment_threshold;
        b.draw_order_threshold = e.draw_order_threshold; b.looping = e.looping; b.hold = e.hold; b.additive = e.additive;
    });
}
void RuntimePlayer::set_track_animation_range(std::uint32_t track, std::optional<std::pair<float, float>> range) {
    current_edit(impl(), "setTrackAnimationRange", track, [&](Playback& p) { p.set_range(track, range); }, [](TrackEntry& b, const TrackEntry& e) { b.start = e.start; b.end = e.end; });
}
void RuntimePlayer::set_track_end(std::uint32_t track, std::optional<float> time) {
    current_edit(impl(), "setTrackEnd", track, [&](Playback& p) { p.set_track_end(track, time); }, [](TrackEntry& b, const TrackEntry& e) { b.track_end = e.track_end; });
}
void RuntimePlayer::set_track_mix_duration(std::uint32_t track, float seconds) {
    current_edit(impl(), "setTrackMixDuration", track, [&](Playback& p) { p.set_mix_duration(track, seconds); }, [](TrackEntry& b, const TrackEntry& e) { b.mix_duration = e.mix_duration; });
}
void RuntimePlayer::set_queued_entry_options(std::uint32_t track, std::size_t index, const QueuedEntryOptions& options) {
    impl().configure("setQueuedEntryOptions", [&](PlayerState& s, EventSink&, EventSink&) {
        s.live.set_queued_options(track, index, options); const auto& e = s.live.tracks().at(track).queue.at(index);
        auto& b = s.baseline.find_entry(track, e.identity, "setQueuedEntryOptions"); b.delay = e.delay; b.mix_duration = e.mix_duration; b.looping = e.looping;
    });
}
void RuntimePlayer::remove_queued_entry(std::uint32_t track, std::size_t index) {
    impl().configure("removeQueuedEntry", [&](PlayerState& s, EventSink& live, EventSink& baseline) {
        (void)s.live.query(track); const auto at = s.live.tracks().find(track);
        if (at == s.live.tracks().end() || index >= at->second.queue.size()) throw Error(ErrorCode::not_found, "removeQueuedEntry", "Queued entry does not exist.", "queueIndex");
        const auto identity = at->second.queue[index].identity;
        s.live.remove_queued(track, index, live); s.baseline.remove_queued_identity(track, identity, baseline);
    });
}
std::optional<TrackState> RuntimePlayer::query_track(std::uint32_t track) const { return impl().state->live.query(track); }
std::vector<QueuedEntryState> RuntimePlayer::query_queued_entries(std::uint32_t track) const { return impl().state->live.queued(track); }
std::uint32_t RuntimePlayer::track_count() const { return impl().state->live.allocated_tracks(); }
float RuntimePlayer::default_mix() const { return impl().state->live.default_mix(); }
}
